include/xplat/Starboard/SmartTypes.h

```diff
@@ -305,16 +305,250 @@ bool operator!=(const Any& other, const AutoId<TObj, TLifetimeTraits>& val) {
+    // AutoCF<T> x, y; x = y;
+    // thus breaking the refcounting on the contents of x.
+    //
+    // The cross-lifetime copy assignment operator and constructor allow for a retaining AutoCF to take a +1
+    // refcount on another weakly-held AutoCF's value.
+    template <typename TOtherLifetime>
+    AutoCF<T, TLifetimeTraits>& operator=(const AutoCF<T, TOtherLifetime>& other) {
+        TLifetimeTraits::store(_addressof(), static_cast<CFTypeRef>(other._val));
+        return *this;
+    }
+
+    AutoCF<T, TLifetimeTraits>& operator=(const AutoCF<T, TLifetimeTraits>& other) {
+        TLifetimeTraits::store(_addressof(), static_cast<CFTypeRef>(other._val));
+        return *this;
+    }
+
+    // The move constructor and assignment operator disallow moving across lifetimes for reasons similar to the
+    // copy allowing them: It isn't reasonable to inherit a +0 refcount in a refcounting container, and neither is
+    // the converse.
+    template <typename TOtherLifetime>
+    AutoCF<T, TLifetimeTraits>& operator=(AutoCF<T, TOtherLifetime>&& other) {
+        static_assert(std::is_same<TLifetimeTraits, TOtherLifetime>::value,
+                      "AutoCF can only move from another AutoCF&& with the same lifetime.");
+        return *this;
+    }
+
+    AutoCF<T, TLifetimeTraits>& operator=(AutoCF<T, TLifetimeTraits>&& other) {
+        _val = other._val;
+        other._val = nullptr;
+        return *this;
+    }
+
+    template <typename Any, typename = typename std::enable_if<TCanConvertFrom<Any>::value>::type>
+    AutoCF<T>& operator=(const Any& val) {
+        TLifetimeTraits::store(_addressof(), static_cast<CFTypeRef>(val));
+        return *this;
+    }
+
+    template <typename Any>
+    bool operator!=(const Any& other) const {
+        // Use a C-style cast here because we want static_cast behaviour for convertible types and
+        // reinterpret_cast behaviour for pointer/value types that can't be converted.
+        return (T)other != _val;
+    }
+
+    template <typename Any>
+    bool operator==(const Any& other) const {
+        // As above.
+        return (T)other == _val;
+    }
+
+    T operator->() {
+        return _val;
+    }
+
+    T& get() {
+        return _val;
+    }
+
+    T get() const {
+        return _val;
+    }
+
+    details::AutoCFRef<AutoCF<T, TLifetimeTraits>, T> operator&() {
+        return details::AutoCFRef<AutoCF<T, TLifetimeTraits>, T>(this);
+    }
+
+    template <typename TOtherObj, typename TOtherLifetime>
+    friend class AutoCF;
+};
+
 template <typename T>
-class unique_cf : public std::unique_ptr<typename std::remove_pointer<T>::type, decltype(&CFRelease)> {
+class unique_cf: public AutoCF<T, CFLifetimeRetain> {
 public:
-    unique_cf() : std::unique_ptr<typename std::remove_pointer<T>::type, decltype(&CFRelease)>(nullptr, CFRelease) {
+    unique_cf(): AutoCF<T, CFLifetimeRetain>() {
+    }
+
+    // unique_cf used to have inherit semantics on its default constructor.
```

Comment:
Using the past tense since it's a "compatibility shim" now, and i'm documenting its prior/intended behaviour. I can clear that up if necessary :)